core/src/main/scala/kafka/log/Log.scala

```diff
@@ -333,42 +332,45 @@ class Log(val dir: File,
 
         if (assignOffsets) {
           // assign offsets to the message set
-          val offset = new AtomicLong(nextOffsetMetadata.messageOffset)
-          val now = SystemTime.milliseconds
-          try {
-            validMessages = validMessages.validateMessagesAndAssignOffsets(offset,
-                                                                           now,
-                                                                           appendInfo.sourceCodec,
-                                                                           appendInfo.targetCodec,
-                                                                           config.compact,
-                                                                           config.messageFormatVersion,
-                                                                           config.messageTimestampType,
-                                                                           config.messageTimestampDifferenceMaxMs)
+          val offset = new LongRef(nextOffsetMetadata.messageOffset)
+          val now = time.milliseconds
+          val (validatedMessages, messagesRecompressed) = try {
+            validMessages.validateMessagesAndAssignOffsets(offset,
+                                                           now,
+                                                           appendInfo.sourceCodec,
+                                                           appendInfo.targetCodec,
+                                                           config.compact,
+                                                           config.messageFormatVersion,
+                                                           config.messageTimestampType,
+                                                           config.messageTimestampDifferenceMaxMs)
           } catch {
             case e: IOException => throw new KafkaException("Error in validating messages while appending to log '%s'".format(name), e)
           }
-          appendInfo.lastOffset = offset.get - 1
-          // If log append time is used, we put the timestamp assigned to the messages in the append info.
+          validMessages = validatedMessages
+          appendInfo.lastOffset = offset.value - 1
           if (config.messageTimestampType == TimestampType.LOG_APPEND_TIME)
             appendInfo.timestamp = now
+
+          if (messagesRecompressed) {
+            // re-validate message sizes since after re-compression some may exceed the limit
```

Comment:
Yes, good point. Then, I think the follower's case is covered. You can leave the code as it is in the patch.